examples/tcp-echo/client.rs

```diff
@@ -0,0 +1,305 @@
+        // Close all connections.
+        for (qd, _) in self.clients.drain().collect::<Vec<_>>() {
+            self.handle_close(qd)?;
+        }
+
+        Ok(())
+    }
+
+    // Makes a scatter-gather array.
+    fn mksga(&mut self, size: usize, value: u8) -> demi_sgarray_t {
+        // Allocate scatter-gather array.
+        let sga: demi_sgarray_t = match self.libos.sgaalloc(size) {
+            Ok(sga) => sga,
+            Err(e) => panic!("failed to allocate scatter-gather array: {:?}", e),
+        };
+
+        // Ensure that scatter-gather array has the requested size.
+        assert!(sga.sga_segs[0].sgaseg_len as usize == size);
+
+        // Fill in scatter-gather array.
+        let ptr: *mut u8 = sga.sga_segs[0].sgaseg_buf as *mut u8;
+        let len: usize = sga.sga_segs[0].sgaseg_len as usize;
+        let slice: &mut [u8] = unsafe { slice::from_raw_parts_mut(ptr, len) };
+        slice.fill(value);
+
+        sga
+    }
+
+    /// Handles the completion of a pop operation.
+    fn handle_pop(&mut self, qr: &demi_qresult_t) -> Result<()> {
+        let qd: QDesc = qr.qr_qd.into();
+        let sga: demi_sgarray_t = unsafe { qr.qr_value.sga };
+        if sga.sga_segs[0].sgaseg_len == 0 {
+            eprint!("INFO: server closed connection");
+            self.handle_close(qd)?;
+        } else {
+            // Retrieve client buffer.
+            let (recvbuf, index): &mut (Vec<u8>, usize) = self
+                .clients
+                .get_mut(&qd)
+                .ok_or(anyhow::anyhow!("unregistered socket"))?;
+
+            // Copy data.
+            let ptr: *mut u8 = sga.sga_segs[0].sgaseg_buf as *mut u8;
+            let len: usize = sga.sga_segs[0].sgaseg_len as usize;
+            let slice: &mut [u8] = unsafe { slice::from_raw_parts_mut(ptr, len) };
+            recvbuf[*index..(*index + len)].copy_from_slice(slice);
+
+            // TODO: Sanity check packet.
+
+            // Free scatter-gather-array.
+            self.libos.sgafree(sga)?;
+
+            *index += len;
+            self.nbytes += len;
+
+            // Check if there are more bytes to read from this packet.
+            if *index < recvbuf.capacity() {
+                // There are, thus issue a partial pop.
+                let size: usize = recvbuf.capacity() - *index;
+                self.issue_pop(qd, Some(size))?;
+            }
+            // Push another packet.
+            else {
+                // There aren't, so push another packet.
+                *index = 0;
+                self.nechoed += 1;
+                self.issue_push(qd)?;
+            }
+        }
+        Ok(())
+    }
+
+    /// Handles the completion of a push operation.
+    fn handle_push(&mut self, qr: &demi_qresult_t) -> Result<()> {
+        let qd: QDesc = qr.qr_qd.into();
+        self.npushed += 1;
+
+        // Pop another packet.
+        self.issue_pop(qd, None)?;
+        Ok(())
+    }
+
+    /// Handles the completion of an expected operation.
```

Comment:
This comment should say: unexpected operation

Reply:
Thanks.